File: banda/SergioDeIscar-banda-java/SergioDeIscar-banda-java/src/test/java/BandaTest.java
import org.example.enums.GuitarraType;
import org.example.enums.PercussionType;
import org.example.enums.TonoType;
import org.example.models.*;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import static org.example.BandaMain.*;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class BandaTest {
    Musician[] musicians = {
        new Bajista("Pepe", 15, 4),
        new Bajista("Pepe2", 15, 4),
        new Cantante("Pepo", 15, TonoType.BARITONO),
        new CantanteGuitarrista("Pepa", 15, TonoType.BARITONO, GuitarraType.FLAMENCA),
        new Guitarrista("Pepi", 15, GuitarraType.FLAMENCA),
        new Multinstrumentista("Pepu", 15, 4,
                PercussionType.ALTURA_INDETERMINADA, 4),
        new Percusionista("Pepo", 15, PercussionType.ALTURA_INDETERMINADA),
        new Teclista("Pepa", 15, 4),
        new Trompetista("Pepi", 15, 4)
    };

    @Test
    public void countType_Test() {
        assertAll(
            () -> assertEquals(2, countType(musicians, Bajista.class)),
            () -> assertEquals(1, countType(musicians, Cantante.class)),
            () -> assertEquals(1, countType(musicians, CantanteGuitarrista.class)),
            () -> assertEquals(1, countType(musicians, Guitarrista.class)),
            () -> assertEquals(1, countType(musicians, Multinstrumentista.class)),
            () -> assertEquals(1, countType(musicians, Percusionista.class)),
            () -> assertEquals(1, countType(musicians, Teclista.class)),
            () -> assertEquals(1, countType(musicians, Trompetista.class))
        );
    }

    @Test
    public void findFist_Test() {
        assertAll(
            () -> assertEquals("Pepe", findFirstType(musicians, Bajista.class).getNombre()),
            () -> assertEquals("Pepo", findFirstType(musicians, Cantante.class).getNombre()),
            () -> assertEquals("Pepa", findFirstType(musicians, CantanteGuitarrista.class).getNombre()),
            () -> assertEquals("Pepi", findFirstType(musicians, Guitarrista.class).getNombre()),
            () -> assertEquals("Pepu", findFirstType(musicians, Multinstrumentista.class).getNombre()),
            () -> assertEquals("Pepo", findFirstType(musicians, Percusionista.class).getNombre()),
            () -> assertEquals("Pepa", findFirstType(musicians, Teclista.class).getNombre()),
            () -> assertEquals("Pepi", findFirstType(musicians, Trompetista.class).getNombre())
        );
    }

    @Test
    public void calculateNomina_Test(){
        float expected = (1500.0f * 1.3f) * 2 +    // Bajista
                1500.0f * 1.4f +                        // Cantante
                1500.0f * 1.5f +                        // CantanteGuitarist
                1500.0f * 1.35f +                       // Guitarrista
                1500.0f * 1.45f +                       // Multinstrumentista
                1500.0f * 1.35f +                       // Percusionista
                1500.0f * 1.3f +                        // Teclista
                1500.0f * 1.2f;                         // Trompetista
        assertEquals(expected, calculateNomina(musicians));
    }
}

File: banda/SergioDeIscar-banda-kotlin/SergioDeIscar-banda-kotlin/src/test/kotlin/BandaTest.kt
import bandamusical.calculateNomina
import bandamusical.countType
import bandamusical.findFirstType
import bandamusical.models.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.assertAll
import kotlin.test.assertEquals

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class BandaTest {
    private val musicians = arrayOf(
        Bajista("pepe", 15, 4),
        Bajista("pepe2", 15, 4),
        Cantante("pepo", 15, TonoType.TENOR),
        CantanteGuitarrista("pepa", 12, TonoType.BARITONO, GuitarraType.CLASICA),
        Guitarrista("pepi", 2, GuitarraType.FLAMENCA),
        Multinstrumentista("pepu", 25, 6, PercussionType.ALTURA_DETERMINADA, 4),
        Percusionista("repepe", 23, PercussionType.ALTURA_INDETERMINADA),
        Teclista("repepo", 7, 3),
        Trompetista("repopa", 9, 0..5)
    )

    @Test
    fun countType_Test() {
        assertAll(
            { assertEquals(2, countType<Bajista>(musicians)) },
            { assertEquals(1, countType<Cantante>(musicians)) },
            { assertEquals(1, countType<Multinstrumentista>(musicians)) }
        )
    }

    @Test
    fun findFirstType_Test(){
        val targetCantante = Cantante("target", 8, TonoType.SOPRANO)
        val targetTeclista = Teclista("target", 8, 7)
        val personasTarget: Array<Musician> = arrayOf(
            targetCantante,
            Cantante("pepe", 10, TonoType.CONTRALTO),
            Cantante("pepe", 10, TonoType.BAJO),
            Cantante("pepe", 10, TonoType.SOPRANO),
            targetTeclista,
            Teclista("pepa", 10, 8),
            Teclista("pepa", 10, 8),
            Teclista("pepa", 10, 8),
            Teclista("pepa", 10, 8)
        )

        assertAll(
            { assertEquals(targetCantante, findFirstType<Cantante>(personasTarget)) },
            { assertEquals(targetTeclista, findFirstType<Teclista>(personasTarget)) },
            { assertEquals(targetCantante, findFirstType<Musician>(personasTarget)) }
        )
    }

    @Test
    fun calculateNomina_Test(){
        val expected = (1500.0f * 1.3f) * 2 +    // Bajista
                1500.0f * 1.4f +                        // Cantante
                1500.0f * 1.5f +                        // CantanteGuitarist
                1500.0f * 1.35f +                       // Guitarrista
                1500.0f * 1.45f +                       // Multinstrumentista
                1500.0f * 1.35f +                       // Percusionista
                1500.0f * 1.3f +                        // Teclista
                1500.0f * 1.2f                          // Trompetista
        assertEquals(expected, calculateNomina(musicians))
    }
}